File: src/swre/swre_interface.ss/java/src/safir/application/Tracer.java
/******************************************************************************
*
* Copyright Saab AB, 2007-2008 (http://www.safirsdk.com)
*
* Created by: Lars Hagström / stlrha
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/
package safir.application;


public class Tracer 
    extends java.io.PrintWriter
{
    public Tracer(String prefix)
    {
        super(new java.io.StringWriter(),true);
        try
        {
            m_traceStream = new TraceStream(prefix);
            super.out = new java.io.BufferedWriter(new java.io.OutputStreamWriter(m_traceStream,"UTF-8"));
        }
        catch (java.io.UnsupportedEncodingException exc)
        {
            //UTF-8 is guaranteed to be supported by the language.
        }
    }
    
    static public void setProgramName(String programName)                                     
    {
        boolean [] success = new boolean [1];
        
        Library.SetProgramName(programName, success);
        if (!success[0])
        {
            safir.dob.typesystem.LibraryExceptions.getInstance().throwFundamental();
            safir.dob.typesystem.LibraryExceptions.getInstance().throwUnknown();
        }
    }

    public boolean isEnabled()
    {
        return Library.TracePrefixIsEnabled(m_traceStream.getPrefixId());
    }
    
    public void setEnabled(boolean enabled)
    {
        boolean [] success = new boolean [1];
        Library.TracePrefixSetEnabled(m_traceStream.getPrefixId(),
                                      enabled,
                                      success);
        if (!success[0])
        {
            safir.dob.typesystem.LibraryExceptions.getInstance().throwFundamental();
            safir.dob.typesystem.LibraryExceptions.getInstance().throwUnknown();
        }
    }

    public java.io.PrintWriter append(char c)
    {
        if (isEnabled())
        {
            super.append(c);
        }
        return this;
    }
    
    public java.io.PrintWriter append(CharSequence csq)
    {
        if (isEnabled())
        {
            super.append(csq);
        }
        return this;
    }

    public java.io.PrintWriter append(CharSequence csq, int start, int end)
    {
        if (isEnabled())
        {
            super.append(csq,start,end);
        }
        return this;
    }
    /*
      public void close()
      {
      if (isEnabled())
      {
            
      }
      }*/
    
      public void flush()
      {
          if (isEnabled())
          {
              boolean [] success = new boolean [1];
              Library.TraceFlushBuffer(success);
              if (!success[0])
              {
                  safir.dob.typesystem.LibraryExceptions.getInstance().throwFundamental();
                  safir.dob.typesystem.LibraryExceptions.getInstance().throwUnknown();
              }
          }
      }

    public java.io.PrintWriter format(java.util.Locale l, String format, Object... args)
    {
        if (isEnabled())
        {
            super.format(l,format,args);
        }
        return this;
    }

    public java.io.PrintWriter format(String format, Object... args)
    {
        if (isEnabled())
        {
            super.format(format,args);
        }
        return this;
    }

    public void print(boolean b)
    {
        if (isEnabled())
        {
            super.print(b);
        }
    }

    public void print(char c)
    {
        if (isEnabled())
        {
            super.print(c);
        }
    }

    public void print(char[] s)
    {
        if (isEnabled())
        {
            super.print(s);
        }
    }

    public void print(double d)
    {
        if (isEnabled())
        {
            super.print(d);
        }
    }

    public void print(float f)
    {
        if (isEnabled())
        {
            super.print(f);
        }
    }

    public void print(int i)
    {
        if (isEnabled())
        {
            super.print(i);
        }
    }

    public void print(long l)
    {
        if (isEnabled())
        {
            super.print(l);
        }
    }

    public void print(Object obj)
    {
        if (isEnabled())
        {
            super.print(obj);
        }
    }

    public void print(String s)
    {
        if (isEnabled())
        {
            super.print(s);
        }
    }

    public java.io.PrintWriter printf(java.util.Locale l, String format, Object... args)
    {
        if (isEnabled())
        {
            super.printf(l,format,args);
        }
        return this;
    }

    public java.io.PrintWriter printf(String format, Object... args)
    {
        if (isEnabled())
        {
            super.printf(format,args);
        }
        return this;
    }

    public void println()
    {
        if (isEnabled())
        {
            super.println();
        }
    }

    public void println(boolean x)
    {
        if (isEnabled())
        {
            super.println(x);
        }
    }

    public void println(char x)
    {
        if (isEnabled())
        {
            super.println(x);
        }
    }

    public void println(char[] x)
    {
        if (isEnabled())
        {
            super.println(x);
        }
    }

    public void println(double x)
    {
        if (isEnabled())
        {
            super.println(x);
        }
    }

    public void println(float x)
    {
        if (isEnabled())
        {
            super.println(x);
        }
    }

    public void println(int x)
    {
        if (isEnabled())
        {
            super.println(x);
        }
    }

    public void println(long x)
    {
        if (isEnabled())
        {
            super.println(x);
        }
    }

    public void println(Object x)
    {
        if (isEnabled())
        {
            super.println(x);
        }
    }

    public void println(String x)
    {
        if (isEnabled())
        {
            super.println(x);
        }
    }

    public void write(char[] buf)
    {
        if (isEnabled())
        {
            super.write(buf);
        }
    }

    public void write(char[] buf, int off, int len)
    {
        if (isEnabled())
        {
            super.write(buf,off,len);
        }
    }

    public void write(int c)
    {
        if (isEnabled())
        {
            super.write(c);
        }
    }

    public void write(String s)
    {
        if (isEnabled())
        {
            super.write(s);
        }
    }

    public void write(String s, int off, int len)
    {
        if (isEnabled())
        {
            super.write(s,off,len);
        }
    }

    

    private TraceStream m_traceStream;
}


class TraceStream
    extends java.io.OutputStream
{

    public TraceStream(String prefix)
    {
        m_prefix = prefix;
    }

    public long getPrefixId()
    {
        if (m_prefixId == 0)
        {
            AddPrefix();
        }
        return m_prefixId;
    }
    
    private void AddPrefix()
    {
        boolean [] success = new boolean [1];
        long [] prefixId = new long [1];
        Library.TracePrefixAdd(m_prefix, prefixId, success);
        
       if (!success[0])
        {
            safir.dob.typesystem.LibraryExceptions.getInstance().throwFundamental();
            safir.dob.typesystem.LibraryExceptions.getInstance().throwUnknown();
        }
       m_prefixId = prefixId[0];
    }


    public void close() throws java.io.IOException
    {
        boolean [] success = new boolean [1];
        Library.TraceFlushBuffer(success);
        if (!success[0])
        {
            safir.dob.typesystem.LibraryExceptions.getInstance().throwFundamental();
            safir.dob.typesystem.LibraryExceptions.getInstance().throwUnknown();
        }
        super.close();
    }

    public void flush() throws java.io.IOException
    {
        super.flush();
        boolean [] success = new boolean [1];
        Library.TraceSyncBuffer(success);
        if (!success[0])
        {
            safir.dob.typesystem.LibraryExceptions.getInstance().throwFundamental();
            safir.dob.typesystem.LibraryExceptions.getInstance().throwUnknown();
        }
    }
    public void write(byte[] b)
    {
        boolean [] success = new boolean [1];
        Library.TraceAppendStringPrefix(m_prefixId,b,success);
        if (!success[0])
        {
            safir.dob.typesystem.LibraryExceptions.getInstance().throwFundamental();
            safir.dob.typesystem.LibraryExceptions.getInstance().throwUnknown();
        }
    }
    
    public void write(int b) 
    {
        boolean [] success = new boolean [1];
        Library.TraceAppendCharPrefix(m_prefixId,(byte)(b & 0xFF),success);
        if (!success[0])
        {
            safir.dob.typesystem.LibraryExceptions.getInstance().throwFundamental();
            safir.dob.typesystem.LibraryExceptions.getInstance().throwUnknown();
        }
    }

    private String m_prefix;
    private long m_prefixId = 0;
}

File: src/swre/swre_interface.ss/dotnet/swre_interface_dotnet/Tracer.cs
/******************************************************************************
*
* Copyright Saab AB, 2007-2013 (http://www.safirsdk.com)
*
* Created by: Lars Hagström / stlrha
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/
using System;
using Safir.SwReports;

namespace Safir.Application
{
    /// <summary>
    /// This class just contains two static methods, for starting and stopping the tracers backdoor.
    /// </summary>
    public sealed class TracerBackdoor
    {
        /// <summary>
        /// Start reception of trace on/off commands
        /// <para/>
        /// The given connection must be opened before this method is called.
        /// If the connection is closed the reception of backdoor commands is
        /// stopped. If a new connection is opened this method needs to be called
        /// in order to start reception of backdoor commands.
        /// <para/>
        /// In situations when a connection is regularly closed and reopened,
        /// for instance in the case of context switches, you should consider
        /// using a dedicated connection as parameter.
        /// </summary>
        /// <param name="connection">The connection used for setting up a subscription for
        ///                          backdoor commands.</param>
        public static void Start(Safir.Dob.ConnectionBase connection)
        {
            byte success;
            Library.SwreC_SetProgramName(System.Text.Encoding.UTF8.GetBytes
                                         (System.Diagnostics.Process.GetCurrentProcess().ProcessName) + char.MinValue,
                                         out success);
            if (!Safir.Dob.Typesystem.Internal.InternalOperations.BoolOf(success))
            {
                Safir.Dob.Typesystem.LibraryExceptions.Instance.Throw();
            }

            Safir.Dob.ConnectionAspectMisc misc = new Safir.Dob.ConnectionAspectMisc(connection);

            Library.SwreC_StartTraceBackdoor(System.Text.Encoding.UTF8.GetBytes(misc.GetConnectionNameCommonPart() + char.MinValue),
                                             System.Text.Encoding.UTF8.GetBytes(misc.GetConnectionNameInstancePart() + char.MinValue),
                                             out success);
            if (!Safir.Dob.Typesystem.Internal.InternalOperations.BoolOf(success))
            {
                Safir.Dob.Typesystem.LibraryExceptions.Instance.Throw();
            }
        }

        ///<summary>
        /// Stop reception of trace on/off commands
        ///</summary>
        static void Stop()
        {
            Library.SwreC_StopTraceBackdoor();
        }
    }

    /// <summary>
    /// A class for trace logging.
    /// </summary>
    public class Tracer : System.IO.StreamWriter
    {
        /// <summary>
        /// Create a logger with a certain prefix.
        /// </summary>
        /// <param name="prefix">The prefix for this logger.</param>
        public Tracer(string prefix)
            : base(new TraceStream(prefix),new System.Text.UTF8Encoding(false)) //Omit BOM
        {
            AutoFlush = true;
        }
     
        /// <summary>
        /// Check/Set whether a certain prefix is enabled or not.
        /// </summary>
        public bool Enabled
        {
            get
            {
                return Safir.Dob.Typesystem.Internal.InternalOperations.BoolOf
                    (Library.SwreC_TracePrefixIsEnabled(((TraceStream)base.BaseStream).GetPrefixId()));
            }

            set
            {
                byte success;
                Library.SwreC_TracePrefixSetEnabled(((TraceStream)base.BaseStream).GetPrefixId(),
                                                    Safir.Dob.Typesystem.Internal.InternalOperations.ByteOf(value),
                                                    out success);

                if (!Safir.Dob.Typesystem.Internal.InternalOperations.BoolOf(success))
                {
                    Safir.Dob.Typesystem.LibraryExceptions.Instance.Throw();
                }
            }
        }
        
        /// <summary>
        /// Flush the buffered data to the logging output.
        /// </summary>
        public override void Flush()
        {
            if (Enabled)
            {
                base.Flush();
            }
        }
        
        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(string value)
        {
            if (Enabled)
            {
                base.Write(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(char[] buffer, int index, int count) 
        {
            if (Enabled)
            {
                base.Write(buffer, index, count);
            } 
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(char[] buffer)
        {
            if (Enabled)
            {
                base.Write(buffer);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(char value) 
        { 
            if (Enabled)
            {
                base.Write(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(string format, params object[] arg)
        {
            if (Enabled)
            {
                base.Write(format, arg);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(string format, object arg0, object arg1, object arg2)
        {
            if (Enabled)
            {
                base.Write(format, arg0, arg1, arg2);
            } 
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(string format, object arg0, object arg1)
        {
            if (Enabled)
            {
                base.Write(format, arg0, arg1);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(string format, object arg0)
        {
            if (Enabled)
            {
                base.Write(format, arg0);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(object value) 
        { 
            if (Enabled)
            {
                base.Write(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(decimal value) 
        { 
            if (Enabled)
            {
                base.Write(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(double value) 
        { 
            if (Enabled)
            {
                base.Write(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(float value) 
        { 
            if (Enabled)
            {
                base.Write(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(ulong value) 
        { 
            if (Enabled)
            {
                base.Write(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(long value) 
        { 
            if (Enabled)
            {
                base.Write(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(uint value) 
        { 
            if (Enabled)
            {
                base.Write(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(int value) 
        { 
            if (Enabled)
            {
                base.Write(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void Write(bool value) 
        { 
            if (Enabled)
            {
                base.Write(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(string value)
        {
            if (Enabled)
            {
                base.WriteLine(value);
            }
        }

        
        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(char[] buffer, int index, int count) 
        {
            if (Enabled)
            {
                base.WriteLine(buffer, index, count);
            } 
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(char[] buffer)
        {
            if (Enabled)
            {
                base.WriteLine(buffer);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(char value) 
        { 
            if (Enabled)
            {
                base.WriteLine(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(string format, params object[] arg)
        {
            if (Enabled)
            {
                base.WriteLine(format, arg);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(string format, object arg0, object arg1, object arg2)
        {
            if (Enabled)
            {
                base.WriteLine(format, arg0, arg1, arg2);
            } 
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(string format, object arg0, object arg1)
        {
            if (Enabled)
            {
                base.WriteLine(format, arg0, arg1);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(string format, object arg0)
        {
            if (Enabled)
            {
                base.WriteLine(format, arg0);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(object value) 
        { 
            if (Enabled)
            {
                base.WriteLine(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(decimal value) 
        { 
            if (Enabled)
            {
                base.WriteLine(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(double value) 
        { 
            if (Enabled)
            {
                base.WriteLine(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(float value) 
        { 
            if (Enabled)
            {
                base.WriteLine(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(ulong value) 
        { 
            if (Enabled)
            {
                base.WriteLine(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(long value) 
        { 
            if (Enabled)
            {
                base.WriteLine(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(uint value) 
        { 
            if (Enabled)
            {
                base.WriteLine(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(int value) 
        { 
            if (Enabled)
            {
                base.WriteLine(value);
            }
        }

        /// <summary>
        /// Override from StreamWriter or TextWriter that 
        /// checks the enabled flag before doing output.
        /// </summary>
        public override void WriteLine(bool value) 
        { 
            if (Enabled)
            {
                base.WriteLine(value);
            }
        }
        
        
    }


    internal class TraceStream : System.IO.Stream
    {
        public TraceStream(string prefix)
        {
            m_prefix = prefix;
            AddPrefix();
        }

        ~TraceStream()
        {

        }

        public System.Int64 GetPrefixId()
        {
            return m_prefixId;
        }

        private void AddPrefix()
        {
            byte success;
            
            Library.SwreC_TracePrefixAdd(System.Text.Encoding.UTF8.GetBytes(m_prefix + char.MinValue), out m_prefixId, out success);

            if (!Safir.Dob.Typesystem.Internal.InternalOperations.BoolOf(success))
            {
                Safir.Dob.Typesystem.LibraryExceptions.Instance.Throw();
            }
        }

        public override bool CanRead
        {
            get { return false; }
        }

        public override bool CanSeek
        {
            get { return false; }
        }

        public override bool CanWrite
        {
            get { return true; }
        }

        public override void Flush()
        { //this is the same as sync in C++
            byte success;
            Library.SwreC_TraceFlush(out success);
            if (!Safir.Dob.Typesystem.Internal.InternalOperations.BoolOf(success))
            {
                Safir.Dob.Typesystem.LibraryExceptions.Instance.Throw();
            }
        }

        public override long Length
        {
            get { throw new NotSupportedException("The TraceStream class is not seekable."); }
        }

        public override long Position
        {
            get
            {
                throw new NotSupportedException("The TraceStream class is not seekable.");
            }
            set
            {
                throw new NotSupportedException("The TraceStream class is not seekable.");
            }
        }

        public override int Read(byte[] buffer, int offset, int count)
        {
            throw new NotSupportedException("The TraceStream class is not readable.");
        }

        public override long Seek(long offset, System.IO.SeekOrigin origin)
        {
            throw new NotSupportedException("The TraceStream class is not seekable.");
        }

        public override void SetLength(long value)
        {
            throw new NotSupportedException("The TraceStream class is not seekable.");
        }

        public override void Write(byte[] buffer, int offset, int count)
        {
            byte success;
            //byte[] str = new byte[count + 1];
            //Array.Copy(buffer, offset, str, 0, count);
            //str[count] = 0; //add null termination
            Library.SwreC_TraceAppendSubstring(m_prefixId, buffer, offset, count, out success);
            if (!Safir.Dob.Typesystem.Internal.InternalOperations.BoolOf(success))
            {
                Safir.Dob.Typesystem.LibraryExceptions.Instance.Throw();
            }
        }

      

        #region Private stuff

        string m_prefix;
        System.Int64 m_prefixId = 0;

        #endregion
    }
}
